### PR TITLE
Add support for security question and answer.

### PR DESCRIPTION
**Pull request by hlbraddock**
@frett

### CODE REVIEW

**Comment by frett** [comment]
looks good, going to run integration tests quick then it should be good to merge

**Comment by frett** [comment]
+1

**Comment by hlbraddock** [comment]
@frett I didn't include security answer encryption at this layer, but the encryption algorithm will need to be common to all applications.

**Comment by hlbraddock** [comment]
@frett This will also need to be merged with https://github.com/GlobalTechnology/idm-user-management/tree/migration2

**Comment by hlbraddock** [comment]
@frett now includes support for hashing security answer

File: api/src/main/java/org/ccci/idm/user/util/HashUtility.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+    private static final Integer BCRYPT_WORK_FACTOR = 12;
+
+    public static String getHash(final String string) {
+        return BCrypt.hashpw(string, BCrypt.gensalt(BCRYPT_WORK_FACTOR));
+    }
+
+    public static boolean checkHash(final String string, final String hash) {
+        return BCrypt.checkpw(string, hash);
+    }
```
Comment (frett) [inline api/src/main/java/org/ccci/idm/user/util/HashUtility.java]: I would add in a simple string normalization function that strips leading & trailing whitespace and converts to lowercase to help ease SA matching.

**Comment by frett** [comment]
just the 1 comment about normalization, other than that it looks good.

for future attributes we add I'd like to prefix the names with some common string (e.g. cru or theKey) that way it easy to distinguish attributes we've added from attributes provided by eDirectory.

**Comment by hlbraddock** [comment]
I can prefix these.

**Comment by hlbraddock** [comment]
Security answer normalized and ldap attributes renamed.

File: api/src/main/java/org/ccci/idm/user/User.java [hunk 2]
```diff
@@ -447,6 +457,53 @@ public void setTelephoneNumber(String telephoneNumber) {
...
+
+    /**
+     * Not meant for public use.
+     */
+    public void setSecurityAnswer(final String securityAnswer, boolean hash) {
+        this.securityAnswer = hash ? HashUtility.getHash(normalize(securityAnswer)) : securityAnswer;
+    }
+
+    private String normalize(final String string) {
+        return CharMatcher.WHITESPACE.trimAndCollapseFrom(string, ' ').toLowerCase();
```
Comment (frett) [inline api/src/main/java/org/ccci/idm/user/User.java]: good call on the collapse as well, I didn't realize that existed in `CharMatcher`

Comment (frett) [inline api/src/main/java/org/ccci/idm/user/User.java]: this will currently trigger an NPE if you call `setSecurityAnswer(null)`

**Comment by hlbraddock** [comment]
Avoiding NPE

**Comment by frett** [comment]
there's a potential security issue when setting SA to null and then checking the string "null", adding a test and modifying the User object to handle it slightly differently

**Comment by hlbraddock** [comment]
Actually, I wonder might it better to always return false when checking null or empty string.

**Comment by frett** [comment]
yeah, there's some subtlety in handling null/empty tests when we are normalizing the input. `" "` would qualify as not empty until we normalize, then it's empty.

the null corner case is because of how BCrypt handles encrypting `null`, it is concatenated with other strings, which turns it into the string `"null"`

**Comment by frett** [comment]
I'm going to make a parameterized test to test out the corner cases I can think of

**Comment by frett** [comment]
alright, that should address corner cases

**Comment by hlbraddock** [comment]
+1

**Comment by frett** [comment]
things look good to me, I'll merge in the morning unless there is anything else you can think of

**Comment by hlbraddock** [comment]
Sounds good. And merge to migration2 branch. I'll need that for migration of security q/a.

**Comment by frett** [comment]
just thought of one more minor thing for my use case, I need a way of knowing whether there is a secret answer set for an account or not, I'm going to add a simple `boolean hasSecurityAnswer` method.
